### PR TITLE
refactor: enforce explicit attribute initialization in PreTrainer

### PR DESCRIPTION
**Pull request by TahaZahid05**
### Context
Previously, the `PreTrainer` class initialized several of its instance variables (such as training/validation/test loaders, inference configuration options, and metrics registry) via side-effecting helper methods (e.g. `set_loaders()`, `_setup_inference()`, and `_attach_metrics()`). This pattern made `__init__` harder to read and trace because attributes were created or mutated inside nested helper calls.

### Changes
To improve code clarity and ensure all initializations happen explicitly in the constructor, the following changes were made:
- **Refactored DataLoader Setup**: Renamed and restructured `set_loaders` into `_build_dataloaders()`, which takes the data configuration and returns a tuple of training, validation, and test loaders, without mutating `self` directly.
- **Refactored Inference Setup**:
  - Assigned `inference_every_epochs`, `inference_num_samples`, and `max_inference_new_tokens` directly from `config.trainer` in `__init__`, avoiding redundant fallbacks.
  - Updated `_setup_inference()` to perform validation and return the resolved tokenizer instead of setting instance variables as side effects.
  - Corrected the conditional check for inference setup to use `is not None` to handle `0` epochs training properly.
- **Refactored Metrics Setup**: Updated `_attach_metrics()` to return the dictionary of attached metrics to be explicitly assigned to `self.metrics` in `__init__`.
- **Constructor Cleanup**: Removed pre-declarations of attributes in `__init__` and replaced them with sequential, explicit variable assignments.

### Verification
- **Unit Tests**: Ran all 74 unit tests via `pytest` (all passed).
- **Integration Test**: 
  - Initialized a starter template project using `trainite init my-experiment --model transformer --dataset string-reverse --trainer pretrainer --force --yes`.
  - Ran the generated training script (`python my-experiment/main.py my-experiment/config.yaml`) with inference logging enabled (`inference_every_epochs: 1`), verifying training, validation evaluation, metric calculation, and inference samples output successfully without regressions.

### CODE REVIEW

File: trainite/trainers/pretrainer.py [hunk 1]
```diff
@@ -209,30 +210,34 @@ def _build_loaders_from_ratios(
...
 
-    def set_loaders(self) -> None:
-        if isinstance(self.config.data, DataConfigBase):
-            self.train_loader = self._build_dataloader(self.config.data.train)
-            self.val_loader = (
-                self._build_dataloader(self.config.data.val)
-                if self.config.data.val
-                else None
+    def _build_dataloaders(
+        self, data_config: DataConfigBase | DataWithAutoSplit
```
Comment (aaishwarymishra) [inline trainite/trainers/pretrainer.py]: I think we should just use self.config , instead of unnecessary parameter passing.

File: trainite/trainers/pretrainer.py [hunk 2]
```diff
@@ -305,11 +310,12 @@ def _eval_step(
         logits = self.model(inputs)
         return {"logits": logits, "targets": targets}
 
-    def _attach_metrics(self) -> None:
+    def _attach_metrics(self) -> dict[str, Loss | Accuracy]:
```
Comment (aaishwarymishra) [inline trainite/trainers/pretrainer.py]: We can define return as standard ignite metric or any, so user can easily add his/metric without needing to touch the union.

File: trainite/trainers/pretrainer.py [hunk 3]
```diff
@@ -209,30 +210,34 @@ def _build_loaders_from_ratios(
...
-            self.train_loader = self._build_dataloader(self.config.data.train)
-            self.val_loader = (
-                self._build_dataloader(self.config.data.val)
-                if self.config.data.val
-                else None
+    def _build_dataloaders(
+        self, data_config: DataConfigBase | DataWithAutoSplit
+    ) -> tuple[DataLoader, DataLoader | None, DataLoader | None]:
+        train_loader: DataLoader
+        val_loader: DataLoader | None = None
```
Comment (aaishwarymishra) [inline trainite/trainers/pretrainer.py]: Hmm might need to check in with victor for this.

**Review by aaishwarymishra — APPROVED** [review]
